File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.py
from collections import Counter

class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if not s or not words:
            return []
        
        word_len = len(words[0])
        word_count = len(words)
        total_len = word_len * word_count
        word_freq = Counter(words)
        result = []
        
        for i in range(word_len):
            left = i
            right = i
            current_count = Counter()
            words_found = 0
            
            while right + word_len <= len(s):
                word = s[right : right + word_len]
                right += word_len
                
                if word in word_freq:
                    current_count[word] += 1
                    words_found += 1
                    
                    while current_count[word] > word_freq[word]:
                        left_word = s[left : left + word_len]
                        current_count[left_word] -= 1
                        words_found -= 1
                        left += word_len
                    
                    if words_found == word_count:
                        result.append(left)
                else:
                    current_count.clear()
                    words_found = 0
                    left = right
                    
        return result

File: 30-substring-with-concatenation-of-all-words/substring-with-concatenation-of-all-words.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<Integer>();
        
        // Edge cases
        if (s == null || s.length() == 0 || words == null || words.length == 0) {
            return result;
        }
        
        int wordLen = words[0].length();
        int wordCount = words.length;
        int totalLen = wordLen * wordCount;
        
        if (s.length() < totalLen) {
            return result;
        }
        
        // Count frequencies of each word in the given array
        Map<String, Integer> wordFreq = new HashMap<String, Integer>();
        for (String word : words) {
            wordFreq.put(word, wordFreq.getOrDefault(word, 0) + 1);
        }
        
        // Run sliding window wordLen times for different starting offsets
        for (int i = 0; i < wordLen; i++) {
            int left = i;
            int right = i;
            int count = 0;
            Map<String, Integer> currentFreq = new HashMap<String, Integer>();
            
            while (right + wordLen <= s.length()) {
                // Extract the next word from the right side of the window
                String word = s.substring(right, right + wordLen);
                right += wordLen;
                
                if (wordFreq.containsKey(word)) {
                    currentFreq.put(word, currentFreq.getOrDefault(word, 0) + 1);
                    count++;
                    
                    // If we have more occurrences of the word than allowed, shrink window from left
                    while (currentFreq.get(word) > wordFreq.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        currentFreq.put(leftWord, currentFreq.get(leftWord) - 1);
                        count--;
                        left += wordLen;
                    }
                    
                    // If window matches the target count, record the starting index
                    if (count == wordCount) {
                        result.add(left);
                    }
                } else {
                    // Invalid word encountered: reset window states
                    currentFreq.clear();
                    count = 0;
                    left = right;
                }
            }
        }
        
        return result;
    }
}
